### PR TITLE
[5.1] TUF Code Cleanup

### PR DESCRIPTION
**Pull request by wilsonge**
Various minor reworkings to the TUF code as below

### Summary of Changes
- Removes the extra HTTP Factory - it serves no purpose. Whilst we don't implement PSR-18 yet it shows the type of object to be passed around which is a created HTTP client - not the factory for it: https://www.php-fig.org/psr/psr-18/ 
- Rework constructor args - db is now required (it already existed in one of the two instantiating classes anyhow and we call Factory::getDbo() already in the other), makes application a required parameter (we were statically grabbing it in the class - which shouldn't be the case for new classes).
- Fixes the constructor doc block

### Testing Instructions
Check all TUF functionality from https://github.com/joomla/joomla-cms/pull/42799 continues to work with no changes. New code is slightly easier to maintain and read.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by wilsonge** [comment]
@SniperSister would be good to understand why we're forcing curl to make the http requests rather than the any of the 3 options the server might support

File: libraries/src/Updater/Update.php [hunk 1]
```diff
@@ -539,7 +539,9 @@ public function loadFromTuf(TufMetadata $metadataTable, string $url, $minimumSta
     {
         $tufFetcher = new TufFetcher(
             $metadataTable,
-            $url
+            $url,
+            Factory::getDbo(),
```
Comment (HLeithner) [inline libraries/src/Updater/Update.php]: This method is deprecated

```suggestion
            Factory::getContainer()->get(DatabaseDriver::class),
```

File: libraries/src/Updater/Update.php [hunk 2]
```diff
@@ -539,7 +539,9 @@ public function loadFromTuf(TufMetadata $metadataTable, string $url, $minimumSta
     {
         $tufFetcher = new TufFetcher(
             $metadataTable,
-            $url
+            $url,
+            Factory::getContainer()->get(DatabaseDriver::class),
```
Comment (joomdonation) [inline libraries/src/Updater/Update.php:544]: After that last commit, you are now missing `use Joomla\Database\DatabaseDriver;` statement. Please add it.

File: libraries/src/TUF/TufFetcher.php [hunk 3]
```diff
@@ -63,15 +70,26 @@ class TufFetcher
...
+    ) {
         $this->metadataTable = $metadataTable;
         $this->repositoryUrl = $repositoryUrl;
-        $this->db            = $db ?? Factory::getContainer()->get(DatabaseDriver::class);
-        $this->httpClient    = $httpClient ?? Factory::getContainer()->get(HttpFactoryInterface::class)->getHttp([], 'curl');
+        $this->db            = $db;
+        $this->app           = $app;
+
+        // TODO: Understand why we only accept curl rather than the best server match
+        $this->httpClient    = $httpClient ?? (new HttpFactory())->getHttp([], 'curl');
```
Comment (HLeithner) [inline libraries/src/TUF/TufFetcher.php]: based on the issue https://github.com/joomla/joomla-cms/issues/42973 we have to check why curl is needed and how our other implementations works with tuf.

@SniperSister

Comment (SniperSister) [reply libraries/src/TUF/TufFetcher.php]: It's fine to remove the pining to curl. It's not required anymore.

Comment (joomdonation) [reply libraries/src/TUF/TufFetcher.php]: @SniperSister Could you please try it? Brian mentioned that without Curl enabled, update checks just hang in his comment https://github.com/joomla/joomla-cms/issues/42973#issuecomment-1981351229 and I see the same behavior while testing this PR.

Comment (SniperSister) [reply libraries/src/TUF/TufFetcher.php]: Will check!

File: libraries/src/TUF/TufFetcher.php [hunk 4]
```diff
@@ -63,15 +70,26 @@ class TufFetcher
...
+     * @param   DatabaseDriver           $db             The database driver
+     * @param   CMSApplicationInterface  $app            The application object for sending errors to users
+     * @param   ?Http                    $httpClient     A client for sending Http requests
      */
-    public function __construct(MetadataTable $metadataTable, string $repositoryUrl, DatabaseDriver $db = null, Http $httpClient = null)
-    {
+    public function __construct(
+        MetadataTable $metadataTable,
+        string $repositoryUrl,
+        DatabaseDriver $db,
```
Comment (laoneo) [inline libraries/src/TUF/TufFetcher.php]: Is it possible to use the interface?

File: libraries/src/TUF/TufFetcher.php (hunk 3 above)
Comment (SniperSister) [reply libraries/src/TUF/TufFetcher.php]: See https://github.com/joomla/joomla-cms/pull/43002

Comment (HLeithner) [reply libraries/src/TUF/TufFetcher.php]: #43002 socket http transport which then also works here.
My suggestion is to remove this line and make the `$httpClient` mandatory and add it in call in Update.php

File: libraries/src/TUF/TufFetcher.php (hunk 4 above)
Comment (HLeithner) [reply libraries/src/TUF/TufFetcher.php]: changed

File: libraries/src/TUF/TufFetcher.php (hunk 3 above)
Comment (HLeithner) [reply libraries/src/TUF/TufFetcher.php]: I made httpClient mandatory and prepared to remove the application from the update, I think it's not needed and is only used for enqueueMessage, which might be done by the caller.

**Comment by alikon** [comment]
hopefully i've done it in a right way
but it works  for me now

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 0eda36925d9d273791fc958843fc1b6f6cdc2097<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42900">issues.joomla.org/tracker/joomla-cms/42900</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 0eda36925d9d273791fc958843fc1b6f6cdc2097<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42900">issues.joomla.org/tracker/joomla-cms/42900</a>.</sub>

**Comment by brianteeman** [comment]
Before this PR my test server without curl would fail to check for updates etc
With this PR there was no problem<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42900">issues.joomla.org/tracker/joomla-cms/42900</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42900">issues.joomla.org/tracker/joomla-cms/42900</a>.</sub>

**Comment by richard67** [comment]
@alikon Have you seen that PHPCS fails in Drone?

File: libraries/src/TUF/TufFetcher.php [hunk 5]
```diff
@@ -9,11 +9,11 @@
 
 namespace Joomla\CMS\TUF;
 
-use Joomla\CMS\Factory;
-use Joomla\CMS\Http\HttpFactoryInterface;
+use Joomla\CMS\Application\CMSApplicationInterface;
+use Joomla\CMS\Http\HttpFactory;
```
Comment (richard67) [inline libraries/src/TUF/TufFetcher.php]:
```suggestion
```
Remove unused import to fix PHPCS in Drone.

**Comment by alikon** [comment]
no, cause 
sometimes drone failure  are ..... 😄

**Comment by richard67** [comment]
> no, cause sometimes drone failure are ..... 😄

@alikon You can't check the drone log to see what fails?

**Comment by alikon** [comment]
not on my local

**Comment by richard67** [comment]
> not on my local

@alikon I mean on GitHub with the link beside the test result. The thing is that when PHPCS fails, no integration or system or unit tests are run, so we cannot see if something's wrong with them.

Anyway I've fixed it now.

**Comment by alikon** [comment]
@richard67 i mean sometimes you got drone failing for timoeout and... something else ..,etc
thanks for fixing it

**Comment by LadySolveig** [comment]
Thanks to all involved 🚀

**Pull request merged**